File: 1889-minimum-space-wasted-from-packaging/1889-minimum-space-wasted-from-packaging.js
/**
 * @param {number[]} packages
 * @param {number[][]} boxes
 * @return {number}
 */
var minWastedSpace = function(packages, boxes) {
    packages.sort((a,b) => a - b)
    boxes.forEach(box => box.sort((a,b)=> a - b))
    const prefix = packages.reduce((acc,curr) => {
        acc.push(acc[acc.length -1] + curr)
        return acc
    },[0])
    let ans = Infinity
    for (let box of boxes){
        let start = 0
        let wasted = 0
        for (let size of box){
            let r = bs(packages,start, size)
            wasted += size * (r-start) - (prefix[r]-prefix[start])
            start = r
        }
        if (start == packages.length)
            ans = Math.min(wasted, ans) 
    }
    return ans == Infinity ? -1 : ans % (10**9 + 7)
    
    function bs(arr, start, target){
        let l = start
        let r = arr.length -1
        while (l<r){
            let m = Math.ceil((l+r)/2)
            if (arr[m] <= target)
                l = m
            else 
                r = m - 1
        }
        if(arr[l] <= target)
            return l + 1
        return l
    }
};

File: 1889-minimum-space-wasted-from-packaging/1889-minimum-space-wasted-from-packaging.py
class Solution:
    def minWastedSpace(self, packages: List[int], boxes: List[List[int]]) -> int:
        packages.sort()
        pre = list(accumulate(packages, initial=0))
        for box in boxes:
            box.sort()
        ans = inf
        
        for box in boxes:
            start = 0
            wasted = 0
            if box[-1] < packages[-1]:
                continue
            for size in box:
                r = bisect.bisect(packages,size,start)
                wasted += size * (r-start) - (pre[r]-pre[start])
                start = r
            if start == len(packages):
                ans = min(wasted, ans)
                ans %= 10**9 + 7
        return ans if ans != inf else -1